### PR TITLE
feat: support multiple origin/destination airports in MCP search

### LINKED ISSUE #98: [Feat] Support for multiple cities in From/To pairs in one search?
^ Title. Support the following usecase in the MCP:

Image

### PR DESCRIPTION
**Pull request by tmchow**
## Summary

MCP `search_flights` and `search_dates` tools now accept comma-separated airport codes for origin and destination (e.g., `origin="JFK,LGA"` `destination="LHR,CDG"`). Single codes still work as before.

## Changes

- Updated `FlightSearchParams` and `DateSearchParams` field descriptions to document comma syntax
- Added `_resolve_airports()` helper in `fli/mcp/server.py` to parse comma-separated codes into Airport lists
- Updated `build_flight_segments()` and `build_date_search_segments()` in `fli/core/builders.py` to accept `Airport | list[Airport]` and build multi-airport `FlightSegment` entries
- The underlying `FlightSegment` model already supports `list[list[Airport | int]]` for multi-airport per leg, so no model changes were needed

## Testing

All 165 existing tests pass. Ruff format and lint clean. The Google Flights API accepts multi-airport segments at the protocol level, though response reliability for multi-airport queries may vary.

Closes #98

This contribution was developed with AI assistance (Claude Code).

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)

<h3>Greptile Summary</h3>

This PR extends the MCP `search_flights` and `search_dates` tools to accept comma-separated airport codes for multi-airport origin/destination searches (e.g., `origin="JFK,LGA"`), requiring no changes to the underlying `FlightSegment` model which already supports multiple airports per leg.

<h3>Confidence Score: 5/5</h3>

Safe to merge; all 165 existing tests pass and the multi-airport logic is correct.

The only finding is a P2 edge case where an empty or all-comma input string silently returns an empty list and produces a less descriptive error than before. This is not a production-breaking issue and does not affect valid inputs.

fli/mcp/server.py — _resolve_airports empty-list guard

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| fli/core/builders.py | Widens origin/destination parameters to Airport | list[Airport]; normalizes to list before building FlightSegment entries — straightforward and correct. |
| fli/mcp/server.py | Adds _resolve_airports() helper for comma-separated codes and updates both execute functions to use it; minor edge case with empty-input error messaging (P2). |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Client as MCP Client
    participant Server as server.py
    participant Resolve as _resolve_airports()
    participant Builder as builders.py
    participant API as Google Flights API

    Client->>Server: search_flights(origin="JFK,LGA", destination="LHR,CDG")
    Server->>Resolve: _resolve_airports("JFK,LGA")
    Resolve-->>Server: [Airport.JFK, Airport.LGA]
    Server->>Resolve: _resolve_airports("LHR,CDG")
    Resolve-->>Server: [Airport.LHR, Airport.CDG]
    Server->>Builder: build_flight_segments(origins=[JFK,LGA], destinations=[LHR,CDG])
    Builder-->>Server: FlightSegment(departure=[[JFK,0],[LGA,0]], arrival=[[LHR,0],[CDG,0]])
    Server->>API: POST /search with multi-airport segment
    API-->>Server: flight results
    Server-->>Client: {success: true, flights: [...]}
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: fli/mcp/server.py
Line: 366-368

Comment:
**Empty-list edge case in `_resolve_airports`**

If `codes` is an empty string or only commas/spaces (e.g., `","`), `_resolve_airports` returns `[]` after filtering. This empty list propagates into `build_flight_segments`, producing a `FlightSegment` with `departure_airport=[]`, which fails Pydantic's `validate_airports` and surfaces only as the generic `"Invalid parameter value"` response — a regression from pre-PR behavior where `resolve_airport("")` raised a descriptive `ParseError`. Adding a guard here restores actionable error messages.

```suggestion
def _resolve_airports(codes: str) -> list[Airport]:
    """Resolve one or more comma-separated airport codes."""
    airports = [resolve_airport(code.strip()) for code in codes.split(",") if code.strip()]
    if not airports:
        raise ParseError(f"No valid airport codes found in: '{codes}'")
    return airports
```

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["feat: support multiple origin/destinatio..."](https://github.com/punitarani/fli/commit/06b09115b7e4f0bb635b9d25c30a466ff788a1f1) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=27339102)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: fli/mcp/server.py [hunk 1]
```diff
@@ -369,10 +386,10 @@ def _execute_flight_search(params: FlightSearchParams) -> dict[str, Any]:
         departure_window = params.departure_window or CONFIG.default_departure_window
         time_restrictions = build_time_restrictions(departure_window) if departure_window else None
 
-        # Build flight segments
+        # Build flight segments (pass full lists for multi-airport support)
         segments, trip_type = build_flight_segments(
-            origin=origin,
-            destination=destination,
+            origin=origins,
+            destination=destinations,
```
Comment (punitarani) [inline fli/mcp/server.py:392]: The `FlightSearchParams` and `DateSearchParams` model descriptions were updated to mention comma-separated codes, but the actual `@mcp.tool` function signatures still describe single airport codes:

```python
origin: Annotated[str, Field(description="Departure airport IATA code (e.g., 'JFK')")]
destination: Annotated[str, Field(description="Arrival airport IATA code (e.g., 'LHR')")]
```

These are the descriptions that MCP clients (LLMs) actually see when discovering tools. They should be updated to match the Params models — e.g.:

```suggestion
    origin: Annotated[str, Field(description="Departure airport IATA code(s), comma-separated for multiple (e.g., 'JFK' or 'JFK,LGA')")],
    destination: Annotated[str, Field(description="Arrival airport IATA code(s), comma-separated for multiple (e.g., 'LHR' or 'LHR,CDG')")],
```

The same applies to the `search_dates` function below (lines 485-486).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fli/mcp/server.py [hunk 2]
```diff
@@ -354,12 +363,20 @@ def _serialize_date_result(date_result: Any) -> dict[str, Any]:
 # =============================================================================
 
 
+def _resolve_airports(codes: str) -> list[Airport]:
+    """Resolve one or more comma-separated airport codes."""
+    airports = [resolve_airport(code.strip()) for code in codes.split(",") if code.strip()]
+    if not airports:
+        raise ParseError(f"No valid airport codes found in: '{codes}'")
+    return airports
+
```
Comment (punitarani) [inline fli/mcp/server.py:372]: No tests were added for the new multi-airport functionality. I'd recommend adding tests for:

1. **`_resolve_airports()`** — single code, multiple codes, whitespace handling (`" JFK , LGA "`), empty string, invalid codes mixed with valid ones
2. **`build_flight_segments()` / `build_date_search_segments()`** with list inputs — verify the `FlightSegment` gets `departure_airport=[[JFK, 0], [LGA, 0]]` structure
3. **`_execute_flight_search()` / `_execute_date_search()`** with comma-separated origin/destination — at minimum a unit test showing the params parse correctly into lists

The existing tests confirm backward compatibility, but without multi-airport tests we can't catch regressions in the new code path.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fli/core/builders.py [hunk 3]
```diff
@@ -87,10 +87,14 @@ def build_flight_segments(
...
+    # Normalize to lists for uniform handling
+    origins = origin if isinstance(origin, list) else [origin]
+    destinations = destination if isinstance(destination, list) else [destination]
+
     segments = [
         FlightSegment(
-            departure_airport=[[origin, 0]],
-            arrival_airport=[[destination, 0]],
+            departure_airport=[[apt, 0] for apt in origins],
+            arrival_airport=[[apt, 0] for apt in destinations],
```
Comment (punitarani) [inline fli/core/builders.py:97]: Nit / edge case to be aware of: the `FlightSegment.validate_airports()` validator only compares the **first** departure airport against the **first** arrival airport. With multi-airport searches, a user could have an airport appear in both origin and destination lists (e.g., `origin="JFK,LGA"` `destination="JFK,CDG"`) and the validator would raise `"Departure and arrival airports must be different"` even though the search might be valid overall.

This is a pre-existing limitation in the model, not introduced by this PR. But multi-airport support makes it more likely to trigger. Worth noting for a follow-up — the validator could check that the origin and destination sets are disjoint rather than just comparing first entries.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by punitarani — CHANGES_REQUESTED** [review]
Nice, clean approach — leveraging the existing `FlightSegment` multi-airport support is the right call. Two changes requested before merging:

1. **MCP tool function descriptions are out of sync** — The `@mcp.tool` function signatures for `search_flights` (line 391-392) and `search_dates` (line 485-486) still describe single airport codes. These are the descriptions MCP clients actually see, so they need to reflect the new comma-separated syntax.

2. **No tests for multi-airport paths** — The existing 165 tests confirm backward compatibility, but there are no tests covering `_resolve_airports()`, list-based builder inputs, or comma-separated MCP params. This new code path needs test coverage.

Additionally, one informational note on a pre-existing `FlightSegment` validator edge case that becomes more relevant with multi-airport support (see inline comment on `builders.py`).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by tmchow** [comment]
Both fixed in 5a5e77d.

1. `@mcp.tool` signatures for `search_flights` and `search_dates` now describe comma-separated codes ("e.g., 'JFK' or 'JFK,LGA'") so MCP clients see the same description as the model-level field.
2. Added 20 new unit tests: 12 for `_resolve_airports` covering single/multi codes, whitespace, case, empty/invalid input; 8 for `build_flight_segments` / `build_date_search_segments` with list inputs on both sides including round trips. `uv run pytest tests/core/test_builders.py tests/mcp/test_multi_airport.py` passes 30/30.

For the `validate_airports` nit, agreed that's a pre-existing limitation of the segment model, not specific to this PR. Happy to open a follow-up if you want that tightened.

File: fli/mcp/server.py (hunk 1 above)
Comment (tmchow) [reply fli/mcp/server.py:392]: Updated in 5a5e77d. Both `search_flights` and `search_dates` now carry the comma-separated description on the tool-level `Field` so LLM clients see the multi-airport syntax, matching the `FlightSearchParams`/`DateSearchParams` model descriptions.

File: fli/mcp/server.py (hunk 2 above)
Comment (tmchow) [reply fli/mcp/server.py:372]: Added in 5a5e77d:

- `tests/mcp/test_multi_airport.py` — 12 tests for `_resolve_airports`: single code, multiple codes, whitespace (`" JFK , LGA "`), lowercase/mixed case, empty string, only-commas, only-whitespace, invalid code, valid+invalid mix, order preservation.
- `tests/core/test_builders.py` — 8 tests for `build_flight_segments` and `build_date_search_segments` with list inputs on both sides, plus round trips verifying the `[[apt, 0] for apt in origins]` structure mirrors correctly on the return leg.

Skipped network-backed tests for `_execute_flight_search` since they'd hit Google Flights. Happy to add them if you'd prefer real-search coverage over unit coverage for the segment assembly.

File: fli/core/builders.py (hunk 3 above)
Comment (tmchow) [reply fli/core/builders.py:97]: Noted, agreed this is the `validate_airports` validator's pre-existing behavior. Kept it out of scope for this PR; happy to open a follow-up to validate cross-list (e.g., no airport shared between origin and destination) if you want it tightened.

**Comment by tmchow** [comment]
@punitarani anything else needed on this?

**Comment by tmchow** [comment]
@punitarani anything else needed on this? Would love to get this merged in
